winml/test/scenario/cppwinrt/CustomOps.cpp

```diff
@@ -0,0 +1,727 @@
+#include "testPch.h"
+#include <wil/result.h>
+#include <D3d11_4.h>
+#include <dxgi1_6.h>
+#include "filehelpers.h"
+#include <fstream>
+#include "SqueezeNetValidator.h"
+#include <winrt/Windows.Graphics.Imaging.h>
+#include <winrt/Windows.Media.h>
+#include "winrt/Windows.Storage.h"
+#include <winrt/Windows.Storage.Streams.h>
+#include <MemoryBuffer.h>
+#include <gsl/gsl>
+// #include "dml/api/DirectML.h"
+#include "CustomOperatorProvider.h"
+
+// For custom operator and shape inferencing support
+#include "core/providers/dml/DmlExecutionProvider/inc/MLOperatorAuthor.h"
+#include "core/providers/dml/OperatorAuthorHelper/MLOperatorAuthorHelper.h"
+#include "core/providers/dml/OperatorAuthorHelper/OperatorHelper.h"
+#include "core/providers/dml/OperatorAuthorHelper/OperatorRegistration.h"
+#include "core/graph/constants.h"
+#include "CustomNullOp.h"
+#include <wil/wrl.h>
+
+using namespace winrt;
+using namespace winrt::Windows::AI::MachineLearning;
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::Media;
+using namespace winrt::Windows::Graphics::Imaging;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::Storage::Streams;
+
+class CustomOpsScenarioTest : public ::testing::Test
+{
+protected:
+    CustomOpsScenarioTest() {
+        init_apartment();
+    }
+};
+
+class CustomOpsScenarioGpuTest : public CustomOpsScenarioTest
+{
+};
+
+// Tests that the execution provider correctly fuses operators together when custom ops are involved.
+TEST_F(CustomOpsScenarioGpuTest, CustomOperatorFusion)
```

Comment:
do these tests pass? we did a bunch of refactoring of the custom op stuff and I was worried we might have messed it up

Reply:
They were passing before I rebased on layer_dev, I'll rerun all tests once I get layer_dev bulding

Reply:
They were passing before I rebase on layer_dev, I'll rerun all tests once I get layer_dev building

Reply:
The following fail on latest layer_dev:

```
CustomOpsScenarioGpuTest.CustomOperatorFusion
CustomOpsScenarioTest.CustomKernelWithBuiltInSchema
CustomOpsScenarioTest.CustomKernelWithCustomSchema
ScenarioCppWinrtTest.Scenario20a_LoadBindEval_CustomOperator_CPU
ScenarioCppWinrtTest.Scenario20b_LoadBindEval_ReplacementCustomOperator_CPU
ScenarioCppWinrtGpuTest.MsftQuantizedModels
ScenarioCppWinrtGpuTest.DeviceLostRecovery
ScenarioCppWinrtGpuSkipEdgeCoreTest.Scenario8_SetDeviceSample_D3D11Device
ScenarioCppWinrtGpuSkipEdgeCoreTest.D2DInterop
```

Reply:
thanks. let's fix those after this is in.

---
In reply to: [350494415](https://github.com/microsoft/onnxruntime/pull/2457#discussion_r350494415) [](ancestors = 350494415)